flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/ChangelogStorageMetricGroup.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.fs;
+
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.Histogram;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.metrics.DescriptiveStatisticsHistogram;
+import org.apache.flink.runtime.metrics.groups.ProxyMetricGroup;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * Metrics related to the Changelog Storage used by the Changelog State Backend. Thread-safety is
+ * required because it is used by multiple uploader threads.
+ */
+@ThreadSafe
```

Comment:
What's the problem here? That those metrics can be updated by many different threads? In opposite to the existing metrics, which are not thread safe, but updates are happening from a single thread?

(Our philosophy so far was that we prefer for metrics to be inaccurate rather then risk performance penalties for synchronisation)

Reply:
Exactly, Thread-safety is required because it is used by multiple uploader threads.

Let me clarify first: 
- existing metric **groups** are thread-safe (though they don't have to be, just by convention)
- the metrics they register are usually not (depends on the implementation, e.g. histogram is thread-safe)

Now, the existing metrics are usually updated by one thread and read by another, so in the worst case results will be delayed.
But here, metrics are **updated** by multiple threads; so results could be corrupted if not thread-safe.